File: data-structures-greedy-algorithms-optimized/src/greedy_algorithms.cpp
```cpp
#include <iostream>
#include <vector>
#include <algorithm> // For std::sort
#include <numeric>   // For std::accumulate
#include <queue>     // For std::priority_queue
#include <map>       // For Huffman frequency map
#include <string>    // For Huffman codes

/**
 * @file greedy_algorithms.cpp
 * @brief Implementations of various Greedy Algorithms with detailed comments and complexity analysis.
 */

// --- Problem 1: Activity Selection Problem ---

/**
 * @brief Represents an activity with a start and finish time.
 */
struct Activity {
    int start;
    int finish;
    int id; // Optional: to identify the original activity

    // Constructor for easy initialization
    Activity(int s = 0, int f = 0, int i = 0) : start(s), finish(f), id(i) {}

    // Comparison operator for sorting based on finish time
    bool operator<(const Activity& other) const {
        return finish < other.finish;
    }
};

/**
 * @brief Solves the Activity Selection Problem using a greedy approach.
 *
 * Given a set of activities, each with a start and finish time, select the
 * maximum number of non-overlapping activities that can be performed.
 *
 * @param activities A vector of `Activity` objects.
 * @return A vector of `Activity` objects representing the selected non-overlapping activities.
 *
 * @complexity
 *   - Time: O(N log N) due to sorting the activities by finish time.
 *           The iteration through sorted activities is O(N).
 *   - Space: O(N) if a copy of activities is made for sorting, or O(1) if sorting in-place
 *            and storing selected activities in a new vector. Here, we modify the input vector copy.
 *
 * @approach
 *   1. **Sort by Finish Time**: The core greedy strategy is to always pick the activity
 *      that finishes earliest among the available (compatible) activities. To facilitate
 *      this, sort all activities in ascending order of their finish times.
 *   2. **Select First Activity**: Pick the first activity from the sorted list (it has
 *      the earliest finish time). Add it to the result set.
 *   3. **Iterate and Select**: Iterate through the remaining activities. For each activity,
 *      check if its start time is greater than or equal to the finish time of the
 *      last selected activity. If it is, this activity is compatible; select it and
 *      update the finish time of the last selected activity.
 *   4. **Proof of Optimality (Sketch)**:
 *      Let `A` be an optimal solution and `G` be the greedy solution.
 *      Assume `G` is not optimal. Let `a1` be the first activity chosen by greedy (earliest finish time).
 *      Let `o1` be the first activity chosen by `A` (sorted by finish time).
 *      Since `a1` has the earliest finish time among all activities, `finish(a1) <= finish(o1)`.
 *      If `a1 != o1`, we can replace `o1` with `a1` in `A` to form a new optimal solution `A'`.
 *      `A'` is also valid because `a1` finishes no later than `o1`, meaning `a1` is compatible
 *      with any activity that `o1` was compatible with.
 *      By repeatedly applying this exchange argument, we can show that the greedy choice
 *      can always be part of an optimal solution. The property that makes greedy work here
 *      is the "optimal substructure" and "greedy choice property".
 */
std::vector<Activity> selectActivities(std::vector<Activity> activities) {
    // Edge case: No activities
    if (activities.empty()) {
        return {};
    }

    // Step 1: Sort activities by their finish times
    // O(N log N)
    std::sort(activities.begin(), activities.end());

    std::vector<Activity> selected;

    // Step 2: Select the first activity (which has the earliest finish time)
    // O(1)
    selected.push_back(activities[0]);
    int last_finish_time = activities[0].finish;

    // Step 3: Iterate through the remaining activities and select compatible ones
    // O(N)
    for (size_t i = 1; i < activities.size(); ++i) {
        // If the current activity's start time is greater than or equal to
        // the finish time of the last selected activity, it's non-overlapping.
        if (activities[i].start >= last_finish_time) {
            selected.push_back(activities[i]);
            last_finish_time = activities[i].finish;
        }
    }

    return selected;
}


// --- Problem 2: Fractional Knapsack Problem ---

/**
 * @brief Represents an item with a weight, value, and calculated value-to-weight ratio.
 */
struct Item {
    int weight;
    int value;
    double ratio; // value / weight

    // Constructor
    Item(int w = 0, int v = 0) : weight(w), value(v) {
        if (w > 0) {
            ratio = static_cast<double>(v) / w;
        } else {
            ratio = 0.0; // Handle zero weight items, though problem usually implies positive weight
        }
    }

    // Comparison operator for sorting based on ratio (descending)
    bool operator<(const Item& other) const {
        return ratio > other.ratio; // Sort in descending order of ratio
    }
};

/**
 * @brief Solves the Fractional Knapsack Problem using a greedy approach.
 *
 * Given a set of items, each with a weight and a value, and a knapsack with a
 * maximum capacity, determine the maximum total value that can be obtained
 * by taking fractions of items.
 *
 * @param items A vector of `Item` objects.
 * @param capacity The maximum capacity of the knapsack.
 * @return The maximum total value that can be obtained.
 *
 * @complexity
 *   - Time: O(N log N) due to sorting the items by value-to-weight ratio.
 *           The iteration through sorted items is O(N).
 *   - Space: O(N) if a copy of items is made for sorting, or O(1) if sorting in-place.
 *
 * @approach
 *   1. **Calculate Ratios**: For each item, calculate its value-to-weight ratio (`value / weight`).
 *   2. **Sort by Ratio**: Sort all items in descending order of their value-to-weight ratios.
 *      This is the greedy choice: always pick the item that yields the most value per unit of weight.
 *   3. **Fill Knapsack**: Iterate through the sorted items:
 *      a. If the current item's weight is less than or equal to the remaining knapsack capacity,
 *         take the whole item. Add its value to the total and decrease capacity by its weight.
 *      b. If the current item's weight is greater than the remaining capacity, take a fraction
 *         of the item. The fraction taken is `remaining_capacity / item_weight`. Add the
 *         corresponding fractional value to the total and set remaining capacity to 0.
 *         Then, break, as the knapsack is full.
 *   4. **Proof of Optimality (Sketch)**:
 *      Suppose there is an optimal solution that does not follow the greedy strategy.
 *      This means there must be two items, `A` and `B`, in the optimal solution such that
 *      `ratio(A) < ratio(B)`, but `A` is taken (partially or fully) and `B` is not (or taken less than it could be).
 *      We can improve this solution by replacing a small amount of `A` with an equal weight of `B`.
 *      Since `ratio(B) > ratio(A)`, replacing `A` with `B` for the same weight will increase the total value.
 *      This contradicts the assumption that the original solution was optimal.
 *      This exchange argument shows that the greedy choice (prioritizing higher ratio items)
 *      always leads to an optimal solution for the Fractional Knapsack problem.
 */
double fractionalKnapsack(std::vector<Item> items, int capacity) {
    // Edge case: No items or zero capacity
    if (items.empty() || capacity <= 0) {
        return 0.0;
    }

    // Step 1 & 2: Calculate ratios (done in Item constructor) and sort by ratio (descending)
    // O(N log N)
    std::sort(items.begin(), items.end());

    double total_value = 0.0;
    int current_capacity = capacity;

    // Step 3: Iterate through sorted items and fill knapsack
    // O(N)
    for (const auto& item : items) {
        if (current_capacity <= 0) {
            break; // Knapsack is full
        }

        if (item.weight <= current_capacity) {
            // Take the whole item
            total_value += item.value;
            current_capacity -= item.weight;
        } else {
            // Take a fraction of the item
            double fraction = static_cast<double>(current_capacity) / item.weight;
            total_value += item.value * fraction;
            current_capacity = 0; // Knapsack is now full
        }
    }

    return total_value;
}


// --- Problem 3: Coin Change Problem (Greedy for specific coin sets) ---

/**
 * @brief Solves the Coin Change Problem using a greedy approach.
 *
 * Given a set of coin denominations and an amount, find the minimum number
 * of coins to make up that amount.
 *
 * IMPORTANT NOTE: This greedy strategy (always picking the largest denomination)
 * ONLY WORKS OPTIMALLY for specific coin systems (e.g., standard US currency:
 * 1, 5, 10, 25, 50, 100 cents). It does NOT work for arbitrary coin sets.
 * For arbitrary coin sets, Dynamic Programming (DP) is required.
 *
 * Example where greedy fails: coins = {1, 3, 4}, amount = 6
 * Greedy: {4, 1, 1} -> 3 coins
 * Optimal (DP): {3, 3} -> 2 coins
 *
 * @param coins A vector of available coin denominations (assumed to be sorted descending).
 * @param amount The target amount.
 * @return The minimum number of coins, or -1 if the amount cannot be made.
 *
 * @complexity
 *   - Time: O(M) where M is the number of coin denominations.
 *           If coins are not sorted, an initial sort takes O(M log M).
 *   - Space: O(1)
 *
 * @approach
 *   1. **Sort Denominations**: Ensure the coin denominations are sorted in descending order.
 *      This allows us to prioritize larger coins.
 *   2. **Iterate and Select**: Iterate through the sorted coin denominations. For each coin:
 *      a. While the remaining `amount` is greater than or equal to the current coin's value,
 *         take that coin. Increment the coin count and subtract the coin's value from the `amount`.
 *   3. **Check Result**: After iterating through all coins, if the `amount` is 0,
 *      the total coin count is the minimum. Otherwise, the amount cannot be made with the given coins.
 *
 * @when_greedy_works
 *   A common condition for greedy coin change to be optimal is if, for any two denominations
 *   `c_i` and `c_j` with `c_i < c_j`, and for any amount `x >= c_j`, the optimal way to
 *   make change for `x` does not involve `c_i` if `x - c_j` can be made optimally without `c_i`.
 *   More formally, for canonical coin systems, a common property is that if we replace the largest
 *   coin `C` with `k` coins of value `c_i`, then `k * c_i = C`.
 */
int coinChangeGreedy(std::vector<int> coins, int amount) {
    // Edge cases
    if (amount < 0) return -1;
    if (amount == 0) return 0;
    if (coins.empty()) return -1;

    // Step 1: Sort coins in descending order.
    // If input is guaranteed sorted, this step can be skipped.
    // O(M log M) where M is number of denominations.
    std::sort(coins.rbegin(), coins.rend()); // Reverse iterators for descending sort

    int num_coins = 0;

    // Step 2: Iterate through coins and take as many as possible
    // O(M) in worst case, as inner loop runs at most (amount / smallest_coin) times total.
    for (int coin : coins) {
        while (amount >= coin) {
            amount -= coin;
            num_coins++;
        }
    }

    // Step 3: Check if the exact amount was made
    if (amount == 0) {
        return num_coins;
    } else {
        return -1; // Cannot make the exact amount
    }
}


// --- Problem 4: Huffman Coding ---

/**
 * @brief Node structure for the Huffman Tree.
 */
struct HuffmanNode {
    char data;          // Character for leaf nodes, '\0' for internal nodes
    int frequency;      // Frequency of the character/subtree
    HuffmanNode *left, *right; // Pointers to left and right children

    // Constructor for leaf nodes
    HuffmanNode(char data_char, int freq) : data(data_char), frequency(freq), left(nullptr), right(nullptr) {}

    // Constructor for internal nodes
    HuffmanNode(int freq, HuffmanNode* l, HuffmanNode* r) : data('\0'), frequency(freq), left(l), right(r) {}

    // Destructor to free memory (important for tree structures)
    ~HuffmanNode() {
        delete left;
        delete right;
    }
};

/**
 * @brief Comparison struct for priority queue (min-heap based on frequency).
 */
struct CompareNodes {
    bool operator()(HuffmanNode* a, HuffmanNode* b) {
        return a->frequency > b->frequency; // Min-heap: smallest frequency has highest priority
    }
};

/**
 * @brief Builds the Huffman Tree using a greedy approach.
 *
 * @param frequencies A map where keys are characters and values are their frequencies.
 * @return The root node of the constructed Huffman Tree.
 *
 * @complexity
 *   - Time: O(N log N) where N is the number of unique characters.
 *           Building the priority queue takes O(N). Each of (N-1) merges involves
 *           extracting two min elements (O(log N)) and inserting one new element (O(log N)).
 *   - Space: O(N) for the priority queue and the tree nodes.
 *
 * @approach
 *   1. **Initialize Priority Queue**: Create a min-priority queue and add a `HuffmanNode`
 *      for each character with its frequency. Each node is initially a leaf.
 *   2. **Greedy Merge**: While there is more than one node in the priority queue:
 *      a. Extract the two nodes with the smallest frequencies (greedy choice).
 *      b. Create a new internal node. Its frequency is the sum of the two extracted nodes' frequencies.
 *         The extracted nodes become its left and right children.
 *      c. Insert the new internal node back into the priority queue.
 *   3. **Final Tree**: The last remaining node in the priority queue is the root of the Huffman Tree.
 *
 * @proof_of_optimality (High-level)
 *   Huffman coding's optimality relies on the "greedy choice property" and "optimal substructure."
 *   The greedy choice is to merge the two least frequent nodes. Intuitively, these are the ones
 *   that should have the longest codes, so putting them deepest in the tree (by merging them first)
 *   is optimal.
 *   Formally, it can be proven by showing that there exists an optimal prefix code in which the
 *   two least frequent characters have sibling leaves (are directly merged). Then, by combining
 *   them into a single "meta-character" whose frequency is their sum, the problem reduces to a
 *   smaller, similar problem, demonstrating optimal substructure.
 */
HuffmanNode* buildHuffmanTree(const std::map<char, int>& frequencies) {
    // Edge case: empty frequencies
    if (frequencies.empty()) {
        return nullptr;
    }
    // Edge case: single character. Tree will just be that node.
    if (frequencies.size() == 1) {
        return new HuffmanNode(frequencies.begin()->first, frequencies.begin()->second);
    }

    // Step 1: Initialize priority queue with leaf nodes for each character
    // O(N) to push N nodes
    std::priority_queue<HuffmanNode*, std::vector<HuffmanNode*>, CompareNodes> pq;
    for (auto const& [character, freq] : frequencies) {
        pq.push(new HuffmanNode(character, freq));
    }

    // Step 2: Greedy Merge
    // O(N log N) for N-1 merges, each with 2 pops and 1 push (log N per operation)
    while (pq.size() > 1) {
        // Extract the two nodes with the smallest frequencies
        HuffmanNode* left = pq.top(); pq.pop();
        HuffmanNode* right = pq.top(); pq.pop();

        // Create a new internal node
        HuffmanNode* newNode = new HuffmanNode(left->frequency + right->frequency, left, right);

        // Insert the new node back into the priority queue
        pq.push(newNode);
    }

    // Step 3: The last remaining node is the root of the Huffman Tree
    return pq.top();
}

/**
 * @brief Generates Huffman codes from a Huffman Tree.
 *
 * @param root The root node of the Huffman Tree.
 * @param current_code The current binary code accumulated during traversal.
 * @param huffman_codes A map to store the generated character-code pairs.
 *
 * @complexity
 *   - Time: O(N) where N is the number of nodes in the Huffman tree (roughly 2 * unique_chars - 1).
 *           Each node is visited once.
 *   - Space: O(H) for recursion stack, where H is the height of the tree. O(N) for the result map.
 */
void generateHuffmanCodes(HuffmanNode* root, std::string current_code, std::map<char, std::string>& huffman_codes) {
    if (!root) {
        return;
    }

    // If it's a leaf node, we've found a character and its code
    if (root->data != '\0') { // Check if it's a leaf node (has a character)
        huffman_codes[root->data] = current_code;
    }

    // Traverse left (add '0') and right (add '1')
    generateHuffmanCodes(root->left, current_code + "0", huffman_codes);
    generateHuffmanCodes(root->right, current_code + "1", huffman_codes);
}

// Public interface for Huffman Coding
std::map<char, std::string> huffmanCoding(const std::map<char, int>& frequencies) {
    HuffmanNode* root = buildHuffmanTree(frequencies);
    std::map<char, std::string> codes;

    if (!root) {
        return codes; // Empty map if no frequencies
    }

    // Special case: Single character input. Its code is "0" by convention, or "" if preferred.
    // My generateHuffmanCodes will return an empty string for a single node tree as it has no left/right children.
    // A single character needs a code. Let's make it "0".
    if (frequencies.size() == 1 && root->data != '\0') {
         codes[root->data] = "0";
    } else {
        generateHuffmanCodes(root, "", codes);
    }

    // Clean up memory
    delete root; // The destructor will recursively delete children

    return codes;
}

```

File: optimization-greedy-algorithms-problems-pro/src/main/java/com/example/greedy/GreedyAlgorithms.java
```java
package com.example.greedy;

import com.example.greedy.models.Activity;
import com.example.greedy.models.Item;
import com.example.greedy.models.Job;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * A comprehensive collection of Greedy Algorithm problems and their optimal solutions.
 * Each method includes detailed comments, time, and space complexity analysis.
 */
public class GreedyAlgorithms {

    /**
     * 1. Activity Selection Problem
     * Given a set of activities, each with a start and finish time, select the maximum number of non-overlapping
     * activities that can be performed by a single person or machine.
     *
     * Greedy Choice: Always select the activity that finishes earliest among the remaining activities that do not
     * conflict with previously selected activities.
     *
     * Why it works (Intuition): Choosing the activity that finishes earliest leaves the maximum amount of time
     * available for subsequent activities, thus maximizing the overall count.
     *
     * @param activities A list of Activity objects, where each activity has a start and finish time.
     * @return A list of selected activities that are non-overlapping and maximum in count.
     *
     * Time Complexity: O(N log N) due to sorting, where N is the number of activities.
     * Space Complexity: O(N) for storing the sorted activities and the result list.
     */
    public List<Activity> selectMaxActivities(List<Activity> activities) {
        if (activities == null || activities.isEmpty()) {
            return Collections.emptyList();
        }

        // 1. Sort activities by their finish times.
        // This is the core greedy step.
        activities.sort(Comparator.comparingInt(a -> a.finish));

        List<Activity> selectedActivities = new ArrayList<>();

        // 2. Select the first activity (which has the earliest finish time).
        selectedActivities.add(activities.get(0));
        Activity lastSelectedActivity = activities.get(0);

        // 3. Iterate through the remaining activities.
        // For each activity, if its start time is greater than or equal to the finish time
        // of the last selected activity, then it does not overlap. Select it.
        for (int i = 1; i < activities.size(); i++) {
            Activity currentActivity = activities.get(i);
            if (currentActivity.start >= lastSelectedActivity.finish) {
                selectedActivities.add(currentActivity);
                lastSelectedActivity = currentActivity;
            }
        }

        return selectedActivities;
    }

    /**
     * 2. Fractional Knapsack Problem
     * Given a set of items, each with a weight and a value, and a knapsack with a maximum
     * weight capacity, determine the maximum total value that can be put into the knapsack.
     * Items can be broken into fractions.
     *
     * Greedy Choice: Always pick the item (or fraction of an item) that has the highest
     * value-to-weight ratio.
     *
     * Why it works (Intuition): To maximize value for a given weight capacity, we prioritize
     * items that give us the most value per unit of weight. Since fractions are allowed,
     * we can fill the knapsack perfectly with the most "dense" value items first.
     *
     * @param items    An array of Item objects, where each item has a weight and a value.
     * @param capacity The maximum weight capacity of the knapsack.
     * @return The maximum total value that can be obtained.
     *
     * Time Complexity: O(N log N) due to sorting, where N is the number of items.
     * Space Complexity: O(N) for storing the sorted items (or O(1) if sorting in-place
     *                  and not considering the input array itself for space complexity).
     */
    public double fractionalKnapsack(Item[] items, int capacity) {
        if (items == null || items.length == 0 || capacity <= 0) {
            return 0.0;
        }

        // 1. Sort items in descending order of their value-to-weight ratio.
        // This is the core greedy step.
        Arrays.sort(items, (a, b) -> Double.compare(b.getRatio(), a.getRatio()));

        double totalValue = 0.0;
        int currentWeight = 0;

        // 2. Iterate through the sorted items.
        for (Item item : items) {
            // If the item can be taken whole, take it.
            if (currentWeight + item.weight <= capacity) {
                currentWeight += item.weight;
                totalValue += item.value;
            } else {
                // If the item cannot be taken whole, take a fraction of it
                // to fill the remaining capacity.
                int remainingCapacity = capacity - currentWeight;
                double fraction = (double) remainingCapacity / item.weight;
                totalValue += item.value * fraction;
                currentWeight = capacity; // Knapsack is full
                break; // No more capacity left
            }
        }

        return totalValue;
    }

    /**
     * 3. Coin Change Problem (Greedy Variant)
     * Given a set of coin denominations (assumed to be a canonical set where greedy works,
     * e.g., US currency: 1, 5, 10, 25, 50, 100) and an amount, find the minimum number of
     * coins required to make up that amount.
     *
     * Greedy Choice: Always pick the largest denomination coin that is less than or equal
     * to the remaining amount.
     *
     * Why it works (Intuition for canonical sets): For canonical coin systems, taking the largest
     * possible coin at each step does not prevent a better solution later because the smaller
     * denominations can efficiently make up any remainder. This is not true for all coin sets
     * (e.g., {1, 3, 4} for amount 6; greedy picks 4, then 1, 1 (3 coins), optimal is 3, 3 (2 coins)).
     *
     * @param denominations An array of available coin denominations, sorted in ascending order.
     *                      (e.g., {1, 5, 10, 25}). Assumed to be canonical.
     * @param amount        The target amount to make change for.
     * @return The minimum number of coins, or -1 if the amount cannot be made.
     *
     * Time Complexity: O(D) where D is the number of denominations. Sorting is not required
     *                  if input is already sorted. If not sorted, O(D log D) to sort.
     * Space Complexity: O(1).
     */
    public int greedyCoinChange(int[] denominations, int amount) {
        if (amount < 0) {
            return -1; // Cannot make change for negative amount
        }
        if (amount == 0) {
            return 0; // 0 coins for 0 amount
        }
        if (denominations == null || denominations.length == 0) {
            return -1; // No coins available
        }

        // 1. Sort denominations in descending order.
        // This is necessary for the greedy choice of picking the largest coin first.
        // A copy is made to avoid modifying the input array if desired.
        int[] sortedDenominations = Arrays.copyOf(denominations, denominations.length);
        Arrays.sort(sortedDenominations); // Sorts ascending
        // Reverse to get descending order: iterate from end later or reverse array
        // For simplicity, we can just iterate from the end of the sorted array.

        int coinCount = 0;
        int remainingAmount = amount;

        // 2. Iterate through denominations from largest to smallest.
        for (int i = sortedDenominations.length - 1; i >= 0; i--) {
            int coin = sortedDenominations[i];
            // While the current coin can be used, use it.
            while (remainingAmount >= coin && coin > 0) { // coin > 0 prevents infinite loop with 0-denom coin
                remainingAmount -= coin;
                coinCount++;
            }
        }

        // If remainingAmount is 0, we successfully made change.
        // Otherwise, it's not possible with the given denominations using this greedy approach.
        // Note: For non-canonical coin sets, a dynamic programming approach might be needed
        //       to guarantee optimality. This greedy solution assumes canonical sets.
        return remainingAmount == 0 ? coinCount : -1;
    }

    /**
     * 4. Minimum Number of Platforms Problem
     * Given arrival and departure times of all trains that reach a railway station,
     * find the minimum number of platforms required for the station so that no train waits.
     * Each platform can only be used by one train at a time.
     *
     * Greedy Choice: At any point in time, we want to know the maximum number of trains
     * simultaneously present at the station. This maximum number will be our answer.
     * We process events (arrivals and departures) in chronological order.
     *
     * Why it works (Intuition): By processing events in sorted order, we can simulate
     * the station's state. An arrival increments the platforms needed, a departure
     * decrements it. The peak value reached during this simulation is the minimum
     * required platforms. If two events happen at the same time, it's better to
     * prioritize departures to free up a platform before a new arrival claims it.
     *
     * @param arrival   An array of arrival times.
     * @param departure An array of departure times.
     * @return The minimum number of platforms required.
     *
     * Time Complexity: O(N log N) due to sorting, where N is the number of trains.
     * Space Complexity: O(N) for storing sorted copies of arrival and departure arrays.
     */
    public int findMinPlatforms(int[] arrival, int[] departure) {
        if (arrival == null || departure == null || arrival.length == 0 || arrival.length != departure.length) {
            return 0; // Or throw an IllegalArgumentException
        }

        int n = arrival.length;

        // 1. Sort both arrival and departure arrays.
        // This allows us to process events in chronological order.
        Arrays.sort(arrival);
        Arrays.sort(departure);

        int platformsNeeded = 0;
        int maxPlatforms = 0;
        int i = 0; // Pointer for arrival array
        int j = 0; // Pointer for departure array

        // 2. Iterate through sorted arrival and departure times using two pointers.
        while (i < n && j < n) {
            // If an arrival event occurs before or at the same time as a departure event,
            // a new platform is needed.
            // If arrival == departure, prioritize departure (j++) to free up platform.
            // This is crucial to get the minimum platforms correctly.
            if (arrival[i] <= departure[j]) {
                platformsNeeded++;
                i++;
            } else { // A departure event occurs first, a platform becomes free.
                platformsNeeded--;
                j++;
            }
            // Update the maximum platforms needed encountered so far.
            maxPlatforms = Math.max(maxPlatforms, platformsNeeded);
        }

        return maxPlatforms;
    }

    /**
     * 5. Job Sequencing with Deadlines Problem
     * Given a set of jobs, each with an ID, a deadline, and a profit, find a sequence
     * of jobs that maximizes the total profit. Each job takes unit time to complete,
     * and can only be done on or before its deadline. Only one machine is available.
     *
     * Greedy Choice: Sort jobs by profit in descending order. Iterate through the sorted
     * jobs and schedule each job in the latest possible time slot (before or on its deadline)
     * that is still available.
     *
     * Why it works (Intuition): Prioritizing higher profit jobs ensures that if a choice
     * has to be made between two jobs for a time slot, the one with higher profit is preferred.
     * Scheduling a job in the latest possible slot leaves earlier slots free for other jobs
     * that might have earlier deadlines but lower profit (which might still be the next
     * best choice to fit).
     *
     * @param jobs An array of Job objects, where each job has an ID, deadline, and profit.
     * @return A list of Job IDs representing the optimal sequence for maximum profit.
     *
     * Time Complexity: O(N log N) due to sorting, plus O(N * max_deadline) in the worst case
     *                  for finding slots if max_deadline is large. If using a Disjoint Set Union
     *                  (DSU) data structure, it can be optimized to O(N log N + N * alpha(max_deadline)),
     *                  where alpha is the inverse Ackermann function, practically constant.
     *                  Here, we use a boolean array for simplicity, which leads to the N*max_deadline term.
     * Space Complexity: O(max_deadline) for the `slots` array, plus O(N) for the result list.
     */
    public List<Character> jobSequencingWithDeadlines(Job[] jobs) {
        if (jobs == null || jobs.length == 0) {
            return Collections.emptyList();
        }

        // 1. Sort jobs by profit in descending order.
        // This is the core greedy step.
        Arrays.sort(jobs, (a, b) -> Integer.compare(b.profit, a.profit));

        // 2. Determine the maximum deadline to initialize the time slots array.
        int maxDeadline = 0;
        for (Job job : jobs) {
            maxDeadline = Math.max(maxDeadline, job.deadline);
        }

        // `slots` array keeps track of occupied time slots.
        // slots[i] = true means time slot `i` is occupied.
        // Index `0` is unused, slots represent time from 1 to maxDeadline.
        boolean[] slots = new boolean[maxDeadline + 1];
        List<Character> scheduledJobs = new ArrayList<>();
        int totalProfit = 0; // Optional: track total profit

        // 3. Iterate through the sorted jobs.
        for (Job job : jobs) {
            // Try to schedule the job in the latest possible free slot before or at its deadline.
            // Iterating downwards from deadline ensures we leave earlier slots free for jobs
            // that might have earlier deadlines.
            for (int k = job.deadline; k >= 1; k--) {
                if (!slots[k]) { // If slot k is free
                    slots[k] = true; // Mark slot as occupied
                    scheduledJobs.add(job.id); // Add job to the scheduled list
                    totalProfit += job.profit; // Add profit
                    break; // Move to the next job
                }
            }
        }
        // System.out.println("Total profit: " + totalProfit); // For debugging/information
        return scheduledJobs;
    }
}
```